File: site/src/commonTest/kotlin/net/numa08/niconico_advertiser_list2/util/VideoIdExtractorTest.kt
package net.numa08.niconico_advertiser_list2.util

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

/**
 * VideoIdExtractorのユニットテスト
 */
class VideoIdExtractorTest {
    // 動画ID直接入力のテスト
    @Test
    fun extractVideoId_withDirectVideoId_returnsVideoId() {
        // 正常系: 動画IDを直接入力
        val input = "sm12345678"
        val result = VideoIdExtractor.extractVideoId(input)
        assertEquals("sm12345678", result)
    }

    @Test
    fun extractVideoId_withDirectVideoIdAndWhitespace_returnsVideoId() {
        // 正常系: 動画IDを前後に空白付きで入力
        val input = "  sm12345678  "
        val result = VideoIdExtractor.extractVideoId(input)
        assertEquals("sm12345678", result)
    }

    @Test
    fun extractVideoId_withDirectVideoIdSingleDigit_returnsVideoId() {
        // 正常系: 1桁の数字の動画ID
        val input = "sm1"
        val result = VideoIdExtractor.extractVideoId(input)
        assertEquals("sm1", result)
    }

    @Test
    fun extractVideoId_withDirectSmOnly_returnsNull() {
        // 異常系: smのみ（動画ID直接入力）
        val input = "sm"
        val result = VideoIdExtractor.extractVideoId(input)
        assertNull(result)
    }

    @Test
    fun extractVideoId_withDirectNonNumericAfterSm_returnsNull() {
        // 異常系: sm以降が数字でない（動画ID直接入力）
        val input = "smabc123"
        val result = VideoIdExtractor.extractVideoId(input)
        assertNull(result)
    }

    // URL形式のテスト
    @Test
    fun extractVideoId_withValidHttpsWwwUrl_returnsVideoId() {
        // 正常系: https://www.nicovideo.jp/watch/sm12345678
        val url = "https://www.nicovideo.jp/watch/sm12345678"
        val result = VideoIdExtractor.extractVideoId(url)
        assertEquals("sm12345678", result)
    }

    @Test
    fun extractVideoId_withValidHttpsUrl_returnsVideoId() {
        // 正常系: https://nicovideo.jp/watch/sm12345678
        val url = "https://nicovideo.jp/watch/sm12345678"
        val result = VideoIdExtractor.extractVideoId(url)
        assertEquals("sm12345678", result)
    }

    @Test
    fun extractVideoId_withValidHttpUrl_returnsVideoId() {
        // 正常系: http://www.nicovideo.jp/watch/sm12345678
        val url = "http://www.nicovideo.jp/watch/sm12345678"
        val result = VideoIdExtractor.extractVideoId(url)
        assertEquals("sm12345678", result)
    }

    @Test
    fun extractVideoId_withSingleDigit_returnsVideoId() {
        // 正常系: 1桁の数字
        val url = "https://www.nicovideo.jp/watch/sm1"
        val result = VideoIdExtractor.extractVideoId(url)
        assertEquals("sm1", result)
    }

    @Test
    fun extractVideoId_withSubdomain_returnsVideoId() {
        // 正常系: サブドメイン付き
        val url = "https://sp.nicovideo.jp/watch/sm12345678"
        val result = VideoIdExtractor.extractVideoId(url)
        assertEquals("sm12345678", result)
    }

    @Test
    fun extractVideoId_withQueryParameter_returnsVideoId() {
        // 正常系: クエリパラメータ付き
        val url = "https://www.nicovideo.jp/watch/sm12345678?ref=top"
        val result = VideoIdExtractor.extractVideoId(url)
        assertEquals("sm12345678", result)
    }

    @Test
    fun extractVideoId_withFragment_returnsVideoId() {
        // 正常系: フラグメント付き
        val url = "https://www.nicovideo.jp/watch/sm12345678#comments"
        val result = VideoIdExtractor.extractVideoId(url)
        assertEquals("sm12345678", result)
    }

    @Test
    fun extractVideoId_withQueryAndFragment_returnsVideoId() {
        // 正常系: クエリパラメータとフラグメント両方
        val url = "https://www.nicovideo.jp/watch/sm12345678?ref=top#comments"
        val result = VideoIdExtractor.extractVideoId(url)
        assertEquals("sm12345678", result)
    }

    @Test
    fun extractVideoId_withEmptyString_returnsNull() {
        // 異常系: 空文字
        val url = ""
        val result = VideoIdExtractor.extractVideoId(url)
        assertNull(result)
    }

    @Test
    fun extractVideoId_withBlankString_returnsNull() {
        // 異常系: 空白文字
        val url = "   "
        val result = VideoIdExtractor.extractVideoId(url)
        assertNull(result)
    }

    @Test
    fun extractVideoId_withoutScheme_returnsNull() {
        // 異常系: スキームなし
        val url = "www.nicovideo.jp/watch/sm12345678"
        val result = VideoIdExtractor.extractVideoId(url)
        assertNull(result)
    }

    @Test
    fun extractVideoId_withDifferentDomain_returnsNull() {
        // 異常系: 違うドメイン
        val url = "https://www.youtube.com/watch/sm12345678"
        val result = VideoIdExtractor.extractVideoId(url)
        assertNull(result)
    }

    @Test
    fun extractVideoId_withoutWatchPath_returnsNull() {
        // 異常系: /watch/パスがない
        val url = "https://www.nicovideo.jp/video/sm12345678"
        val result = VideoIdExtractor.extractVideoId(url)
        assertNull(result)
    }

    @Test
    fun extractVideoId_withoutSmPrefix_returnsNull() {
        // 異常系: smで始まらない
        val url = "https://www.nicovideo.jp/watch/12345678"
        val result = VideoIdExtractor.extractVideoId(url)
        assertNull(result)
    }

    @Test
    fun extractVideoId_withSmOnly_returnsNull() {
        // 異常系: smだけ
        val url = "https://www.nicovideo.jp/watch/sm"
        val result = VideoIdExtractor.extractVideoId(url)
        assertNull(result)
    }

    @Test
    fun extractVideoId_withNonNumericAfterSm_returnsNull() {
        // 異常系: sm以降が数字でない
        val url = "https://www.nicovideo.jp/watch/smabc123"
        val result = VideoIdExtractor.extractVideoId(url)
        assertNull(result)
    }

    @Test
    fun extractVideoId_withMixedCaseSmPrefix_returnsNull() {
        // 異常系: SMが大文字（小文字のsmのみ許可）
        val url = "https://www.nicovideo.jp/watch/SM12345678"
        val result = VideoIdExtractor.extractVideoId(url)
        assertNull(result)
    }

    @Test
    fun extractVideoId_withDomainSuffix_returnsNull() {
        // 異常系: nicovideo.jpで終わるが、正しいドメインではない
        val url = "https://fakenicovideo.jp/watch/sm12345678"
        val result = VideoIdExtractor.extractVideoId(url)
        assertNull(result)
    }
}

File: frontend/test/videoIdExtractor.test.ts
// 対応仕様: docs/FRONTEND_REQUIREMENTS.md
//   FE-120〜FE-124（動画ID抽出）
// テストケースは現行KotlinのVideoIdExtractorTest（受け入れ基準）を移植したもの
import { describe, expect, it } from "vitest";
import { extractVideoId } from "../src/lib/videoIdExtractor";

describe("extractVideoId", () => {
  it("FE-122: 直接の動画IDはそのまま返す", () => {
    expect(extractVideoId("sm12345678")).toBe("sm12345678");
    expect(extractVideoId("sm1")).toBe("sm1");
  });

  it("FE-120: 前後の空白を除去してから判定する", () => {
    expect(extractVideoId("  sm12345678  ")).toBe("sm12345678");
  });

  it("FE-123: nicovideo.jpのwatch URLから動画IDを抽出する", () => {
    expect(extractVideoId("https://www.nicovideo.jp/watch/sm12345678")).toBe("sm12345678");
    expect(extractVideoId("https://nicovideo.jp/watch/sm12345678")).toBe("sm12345678");
    expect(extractVideoId("http://www.nicovideo.jp/watch/sm12345678")).toBe("sm12345678");
    expect(extractVideoId("https://sp.nicovideo.jp/watch/sm12345678")).toBe("sm12345678");
    expect(extractVideoId("https://www.nicovideo.jp/watch/sm1")).toBe("sm1");
  });

  it("FE-123: ホストの判定は大文字小文字を区別しない", () => {
    expect(extractVideoId("https://WWW.NICOVIDEO.JP/watch/sm12345678")).toBe("sm12345678");
  });

  it("FE-123: クエリパラメータ・フラグメントを除いた動画IDを返す", () => {
    expect(extractVideoId("https://www.nicovideo.jp/watch/sm12345678?ref=ranking")).toBe(
      "sm12345678",
    );
    expect(extractVideoId("https://www.nicovideo.jp/watch/sm12345678#comment")).toBe("sm12345678");
    expect(extractVideoId("https://www.nicovideo.jp/watch/sm12345678?ref=ranking#comment")).toBe(
      "sm12345678",
    );
  });

  it("FE-124: 空・空白のみは失敗する", () => {
    expect(extractVideoId("")).toBeNull();
    expect(extractVideoId("   ")).toBeNull();
  });

  it("FE-121/FE-124: 動画ID形式でない直接入力は失敗する", () => {
    expect(extractVideoId("sm")).toBeNull();
    expect(extractVideoId("smabc")).toBeNull();
    expect(extractVideoId("SM12345678")).toBeNull();
    expect(extractVideoId("12345678")).toBeNull();
  });

  it("FE-124: スキームなし・別ドメイン・接尾一致ドメインは失敗する", () => {
    expect(extractVideoId("www.nicovideo.jp/watch/sm12345678")).toBeNull();
    expect(extractVideoId("https://example.com/watch/sm12345678")).toBeNull();
    expect(extractVideoId("https://fakenicovideo.jp/watch/sm12345678")).toBeNull();
  });

  it("FE-124: /watch/以外のパスは失敗する", () => {
    expect(extractVideoId("https://www.nicovideo.jp/user/12345")).toBeNull();
    expect(extractVideoId("https://www.nicovideo.jp/")).toBeNull();
  });

  it("FE-124: URL中の動画IDが不正な場合は失敗する", () => {
    expect(extractVideoId("https://www.nicovideo.jp/watch/sm")).toBeNull();
    expect(extractVideoId("https://www.nicovideo.jp/watch/smabc")).toBeNull();
  });
});
